packages/zk-certificates/contracts/HumanIDSaltRegistry.sol

```diff
@@ -0,0 +1,179 @@
+    uint256[] zkCerts;
+}
+
+/**
+ * @title HumanIDSaltRegistry
+ * @author Galactica dev team
+ * @notice Registry to limit a user to one salt to achieve a unique and single dApp specific humanID.
+ */
+contract HumanIDSaltRegistry {
+    GuardianRegistry public guardianRegistry;
+    address public zkCertRegistry;
+
+    // mapping from  id hash to the salt hash
+    mapping(uint256 => UserData) internal _userData;
+    mapping(uint256 => SaltLockingZkCert) internal _saltLockingZkCerts;
+
+    constructor(address guardianRegistry_, address zkCertRegistry_) {
+        guardianRegistry = GuardianRegistry(guardianRegistry_);
+        zkCertRegistry = zkCertRegistry_;
+    }
+
+    /**
+     * @notice Register a salt hash for a user. Reverts if another salt is registered already. Saves zkCert data for eventual recovery.
+     * @param zkCert - Data for the SaltLockingZkCert.
+     * @param idHash - Hash identifying the user. It is supposed to be the poseidon Hash of the name, birthday and citizenship.
+     * @param saltHash - Hash of the salt.
+     */
+    function onZkCertIssuance(
+        SaltLockingZkCert calldata zkCert,
+        uint256 idHash,
+        uint256 saltHash
+    ) external {
+        require(
+            msg.sender == zkCertRegistry,
+            'HumanIDSaltRegistry: only zkCertRegistry can call this function'
+        );
+
+        _registerSaltHash(idHash, saltHash);
+
+        _userData[idHash].zkCerts.push(zkCert.zkCertId);
+        _saltLockingZkCerts[zkCert.zkCertId] = zkCert;
+    }
+
+    /**
+     * @notice Inform salt registry about zkCert revocation. This helps resetting the salt before zkCert expiration if needed.
+     * @param zkCertId - Id of the zkCert to be revoked.
+     */
+    function onZkCertRevocation(uint256 zkCertId) external {
+        require(
+            msg.sender == zkCertRegistry,
+            'HumanIDSaltRegistry: only zkCertRegistry can call this function'
+        );
+
+        _saltLockingZkCerts[zkCertId].revoked = true;
+    }
+
+    /**
+     * @notice Lists zkCerts locking the user to a salt hash.
+     * @param idHash - Hash identifying the user. It is supposed to be the poseidon Hash of the name, birthday and citizenship.
+     * @return List of zkCerts blocking the reset. If it is [], the salt can be reset.
+     */
+    function getSaltLockingZkCerts(
+        uint256 idHash
+    ) public view returns (SaltLockingZkCert[] memory) {
+        require(
+            guardianRegistry.isWhitelisted(msg.sender),
+            'HumanIDSaltRegistry: only whitelisted guardians can call this function'
+        );
+        // a technically advanced user could still extract the data from the blockchain. This is just to make it harder so that the average user can not get it from the block explorer.
+
+        // count zkCerts that are still locking the salt
+        uint256 blockingZkCertsCount = 0;
+        for (uint256 i = 0; i < _userData[idHash].zkCerts.length; i++) {
```

Comment:
minor gas optimization: we can save the length into a memory variable to prevent storage lookup each iteration of the loop, but to be honest I am not sure if its worth it on Galactica and I forget to do it all the time too :D

Reply:
Thanks for the hint. I'll change it to get used to it :smile: